x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/aggregation/RollupDataExtractorFactory.java

```diff
@@ -0,0 +1,219 @@
+import org.elasticsearch.xpack.core.rollup.job.DateHistogramGroupConfig;
+import org.elasticsearch.xpack.ml.datafeed.extractor.DataExtractorFactory;
+import org.joda.time.DateTimeZone;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.xpack.core.ml.datafeed.extractor.ExtractorUtils.getHistogramAggregation;
+import static org.elasticsearch.xpack.core.ml.datafeed.extractor.ExtractorUtils.getHistogramIntervalMillis;
+import static org.elasticsearch.xpack.core.ml.datafeed.extractor.ExtractorUtils.validateAndGetCalendarInterval;
+
+public class RollupDataExtractorFactory implements DataExtractorFactory {
+
+    private final Client client;
+    private final DatafeedConfig datafeedConfig;
+    private final Job job;
+
+    private RollupDataExtractorFactory(Client client, DatafeedConfig datafeedConfig, Job job) {
+        this.client = Objects.requireNonNull(client);
+        this.datafeedConfig = Objects.requireNonNull(datafeedConfig);
+        this.job = Objects.requireNonNull(job);
+    }
+
+    @Override
+    public DataExtractor newExtractor(long start, long end) {
+        long histogramInterval = datafeedConfig.getHistogramIntervalMillis();
+        AggregationDataExtractorContext dataExtractorContext = new AggregationDataExtractorContext(
+            job.getId(),
+            job.getDataDescription().getTimeField(),
+            job.getAnalysisConfig().analysisFields(),
+            datafeedConfig.getIndices(),
+            datafeedConfig.getTypes(),
+            datafeedConfig.getQuery(),
+            datafeedConfig.getAggregations(),
+            Intervals.alignToCeil(start, histogramInterval),
+            Intervals.alignToFloor(end, histogramInterval),
+            job.getAnalysisConfig().getSummaryCountFieldName().equals(DatafeedConfig.DOC_COUNT),
+            datafeedConfig.getHeaders());
+        return new RollupDataExtractor(client, dataExtractorContext);
+    }
+
+    public static void create(Client client,
+                              DatafeedConfig datafeed,
+                              Job job,
+                              Map<String, RollableIndexCaps> rollupJobsWithCaps,
+                              ActionListener<DataExtractorFactory> listener) {
+
+        final AggregationBuilder datafeedHistogramAggregation = getHistogramAggregation(
+            datafeed.getAggregations().getAggregatorFactories());
+        if ((datafeedHistogramAggregation instanceof DateHistogramAggregationBuilder) == false) {
+            listener.onFailure(
+                new IllegalArgumentException("Rollup requires that the datafeed configuration use a [date_histogram] aggregation," +
+                    " not a [histogram] aggregation over the time field."));
```

Comment:
nit: looks like defense in depth, but if this is really triggered, isn't it in theory possible that the aggregation is any aggregation, so you could generalize the message: `"... not a [" + datafeedHistogramAggregation.getName() + "] aggregation ... "`